src/streams.c

```diff
@@ -81,67 +83,121 @@ static Int READ_COMMAND(Obj *evalResult)
  *
- * If the first entry is `true`, then the second entry is bound to the result of
- * the statement if there was one, and unbound otherwise, and the third entry
- * reflects whether the statement ended in a dual semicolon,
- * if the first entry is `false', an error occurred during the execution of a statement,
- * and no other positions of the list are bound.
+ * If the first entry is `true`, then the second entry is bound to the result
+ * of the statement if there was one, and unbound otherwise, and the third
+ * entry reflects whether the statement ended in a dual semicolon, if the
+ * first entry is `false', an error occurred during the execution of a
+ * statement, and no other positions of the list are bound. If outputFunc
+ * is a function, the fourth entry of the list is outputFunc applied to the
+ * result, if there was one, and if the statement did not end in a dual
+ * semicolon. Furthermore, if outputFunc is not false, the fifth entry
+ * contains a string of all output printed while the corresponding statement
+ * was executed. If outputFunc is false, all intermediate output is discarded.
  *
- * This function is currently used in interactive tools such as the GAP Jupyter kernel to
- * execute cells and is likely to be replaced by a function that can read a single command
- * from a stream without losing the rest of its content.
+ * This function is currently used in interactive tools such as the GAP
+ * Jupyter kernel to execute cells and is likely to be replaced by a function
+ * that can read a single command from a stream without losing the rest of its
+ * content.
  */
-Obj FuncREAD_ALL_COMMANDS( Obj self, Obj stream, Obj echo )
+Obj FuncREAD_ALL_COMMANDS(Obj self, Obj instream, Obj echo, Obj outputFunc)
 {
     ExecStatus status;
-    Int resultCount, resultCapacity;
-    UInt dualSemicolon;
-    Obj result, resultList;
-    Obj evalResult;
+    Int        resultCount, resultCapacity;
+    UInt       dualSemicolon;
+    Obj        result, resultList;
+    Obj        evalResult;
+    Obj        outstream;
+    Obj        outstreamString;
+    Obj        outstreamStringCopy;
+    Obj        outputFunc_return;
+
+    /* try to open the streams */
+    if (!OpenInputStream(instream, echo == True)) {
+        return Fail;
+    }
 
-    /* try to open the stream */
-    if (!OpenInputStream(stream, echo == True)) {
+
+    if (outputFunc != False) {
+        outstreamString = NEW_STRING(0);
+        outstream = DoOperation2Args(ValGVar(GVarName("OutputTextString")),
+                                     outstreamString, True);
+    }
+    else {
+        outstreamString = False;
+        outstream = False;
+    }
+    if (outstream != False && !OpenOutputStream(outstream)) {
+        CloseInput();
         return Fail;
     }
 
     resultCount = 0;
     resultCapacity = 16;
-    resultList = NEW_PLIST( T_PLIST, resultCapacity );
+    resultList = NEW_PLIST(T_PLIST, resultCapacity);
```

Comment:
Optional: I would just get rid of both `resultCount ` and esp. `resultCapacity`l the latter does nothing useful, as GROW_PLIST already grows list "intelligently", by increasing its size by at least 25%. So, just remove `resultCapacity`, hardcode `16` as second argument for `NEW_PLIST` above... then [continued below]